unifi_console/datadog_checks/unifi_console/api.py

```diff
@@ -0,0 +1,107 @@
+import functools
+import json
+from typing import Any, Callable, List, TypeVar, cast
+
+from requests import HTTPError, Timeout
+
+from datadog_checks.unifi_console.client_info import ClientInfo
+from datadog_checks.unifi_console.device_info import DeviceInfo
+from datadog_checks.unifi_console.errors import Unauthorized
+from datadog_checks.unifi_console.types import APIConnectionError, ControllerInfo
+
+CallableT = TypeVar("CallableT", bound=Callable)
+
+
+def smart_retry(f):
+    # type: (Callable) -> CallableT
+    """A function decorated with this `@smart_retry` will trigger a new authentication if it fails. The function
+    will then be retried.
+    This is useful when the integration keeps a semi-healthy connection to the vSphere API"""
+
+    @functools.wraps(f)
+    def wrapper(api_instance, *args, **kwargs):
+        # type: (UnifiAPI, *Any, **Any) -> Any
+        try:
+            return f(api_instance, *args, **kwargs)
+        except Unauthorized as e:
+
+            api_instance.log.debug(
+                "An exception occurred when executing %s: %s. Refreshing the connection to vCenter and retrying",
+                f.__name__,
+                e,
+            )
+            api_instance.connect()
+            return f(api_instance, *args, **kwargs)
+
+        except Exception:
+            raise
+
+    return cast(CallableT, wrapper)
+
+
+class UnifiAPI(object):
+    def __init__(self, config, http, log) -> None:
+        self.config = config
+        self.log = log
+        self.http = http
+
+    def connect(self) -> None:
+        payload = json.dumps({"username": self.config.user, "password": self.config.password})
```

Comment:
No need to pass these, the RequestsWrappers already handles `username` and `password` that is defined in `instances/http` in the config